lib/msf/core/rhosts_walker.rb

```diff
@@ -0,0 +1,328 @@
+    # @yield [Msf::RhostsWalker::Error] Yields only invalid rhost values.
+    def errors(&block)
+      return unless @value
+      return unless block_given?
+
+      parse(@value, @datastore).each do |result|
+        block.call(result) if result.is_a?(Msf::RhostsWalker::Error)
+      end
+
+      nil
+    end
+
+    #
+    # Indicates that the rhosts value is valid and iterable
+    #
+    # @return [Boolean] True if all items are valid, and there are at least some items present to iterate over. False otherwise.
+    def valid?
+      parsed_values = parse(@value, @datastore)
+      parsed_values.all? { |result| result.is_a?(Msf::DataStore) } && parsed_values.count > 0
+    rescue StandardError => e
+      elog('rhosts walker invalid', error: e)
+      false
+    end
+
+    #
+    # Parses the input rhosts string, and yields the possible combinations of datastore values.
+    #
+    # @param value [String] the rhost string
+    # @param datastore [Msf::Datastore] the datastore
+    # @return [Enumerable<Msf::DataStore|StandardError>] The calculated datastore values that can be iterated over for
+    #   enumerating the given rhosts, or the error that occurred when iterating over the input
+    def parse(value, datastore)
+      Enumerator.new do |results|
+        # extract the individual elements from the rhost string, ensuring that
+        # whitespace, strings, escape characters, etc are handled correctly.
+        values = Rex::Parser::Arguments.from_s(value)
+        values.each do |value|
+          if (value =~ %r{^file://(.*)}) || (value =~ /^file:(.*)/)
+            file = Regexp.last_match(1)
+            File.read(file).each_line(chomp: true) do |line|
+              parse(line, datastore).each do |result|
+                results << result
+              end
+            end
+          elsif value =~ /^cidr:(.*)/
+            cidr, child_value = Regexp.last_match(1).split(':', 2)
+            # Validate cidr syntax matches ipv6 '%scope_id/mask_part' or ipv4 '/mask_part'
+            raise InvalidCIDRURIError unless cidr =~ %r{^(%\w+)?/\d{1,3}$}
+
+            # Parse the values, then apply range walker over the result
+            parse(child_value, datastore).each do |result|
+              host_with_cidr = result['RHOSTS'] + cidr
+              Rex::Socket::RangeWalker.new(host_with_cidr).each_ip do |rhost|
+                results << result.merge('RHOSTS' => rhost, 'UNPARSED_RHOSTS' => value)
+              end
+            end
+          elsif value =~ /^(?<schema>\w+):.*/ && SUPPORTED_SCHEMAS.include?(Regexp.last_match(:schema))
+            parse_method = "parse_#{Regexp.last_match(:schema)}_uri"
+            parsed_options = send(parse_method, value, datastore)
+            Rex::Socket::RangeWalker.new(parsed_options['RHOSTS']).each_ip do |ip|
+              results << datastore.merge(
+                parsed_options.merge('RHOSTS' => ip, 'UNPARSED_RHOSTS' => value)
+              )
+            end
+          else
```

Comment:
TODO: Maybe it would be nice to raise if a protocol is used which is entirely unsupported, rather than falling back to existing RangeWalker behavior which will just drop the values entirely. Note that in this scenario, it will still be validated correctly due to `valid?` requiring at least one rhost target